### PR TITLE
FOUR-13038 Fix Login blocked because invalid 2FA settings

### PR DESCRIPTION
**Pull request by caleeli**
## Issue & Reproduction Steps

1. Log in to an instance without Email Server configured
2. Create a new user
3. Go to admin tab
4. Click on settings in sidebar menu
5. Click on Log in option
6. Enable 2FA
7. Select By email option
8. Log in in other browser with new user
9. Fill username
10. Fill password

## Solution
- Add a validation before to enable the email and sms options.

image

image

## How to Test

1. Log in to an instance without Email Server configured
2. Create a new user
3. Go to admin tab
4. Click on settings in sidebar menu
5. Click on Log in option
6. Enable 2FA
7. Select By email option
8. Validation will be triggered

## Related Tickets & Packages
- https://processmaker.atlassian.net/browse/FOUR-13038

## Code Review Checklist
- [x] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [x] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [x] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [x] This solution fixes the bug reported in the original ticket.
- [x] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [x] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [x] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [x] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [x] This ticket conforms to the PRD associated with this part of ProcessMaker.

ci:next
ci:deploy
ci:APP_DEBUG=1
ci:TWILIO_ACCOUNT_SID="ACf3487ba53a2b560c6a50625068fd7bc3"
ci:TWILIO_AUTH_TOKEN="15a5b1955d7ff6cc394caaa1f8b70b41"
ci:TWILIO_ACTIVE_PHONE_NUMBER="+12057511892"
ci:package-savedsearch:feature/FOUR-11530

### CODE REVIEW

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-cd0b46d888.eng.processmaker.net

File: resources/views/auth/2fa/email_test.blade.php [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<!DOCTYPE html>
+<html lang="en" xml:lang="en">
+<head>
+  <title>{{ __("Test email server status")</title>
```
Comment (devmiguelangel) [inline resources/views/auth/2fa/email_test.blade.php:4]: close brackets

```suggestion
  <title>{{ __("Test email server status") }}</title>
```

File: resources/views/auth/2fa/email_test.blade.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+</head>
+<body>
+  <p>{{ __("Test email server status")</p>
```
Comment (devmiguelangel) [inline resources/views/auth/2fa/email_test.blade.php:7]: close brackets

```suggestion
  <p>{{ __("Test email server status") }}</p>
```

File: resources/js/admin/settings/components/SettingCheckboxes.vue [hunk 3]
```diff
@@ -51,6 +55,8 @@ export default {
   props: ['value', 'setting'],
   data() {
     return {
+      message: "",
```
Comment (devmiguelangel) [inline resources/js/admin/settings/components/SettingCheckboxes.vue:58]: use single quotes

```suggestion
      message: '',
```

File: resources/js/admin/settings/components/SettingCheckboxes.vue [hunk 4]
```diff
@@ -103,6 +109,10 @@ export default {
     }
   },
   methods: {
+    onChanged() {
+      this.invalid = false;
+      this.message = "";
```
Comment (devmiguelangel) [inline resources/js/admin/settings/components/SettingCheckboxes.vue:114]: use single quotes

```suggestion
      this.message = '';
```

File: resources/js/admin/settings/components/SettingCheckboxes.vue [hunk 5]
```diff
@@ -112,8 +122,12 @@ export default {
     onModalHidden() {
       this.transformed = this.copy(this.input);
       this.error = false;
-      this.loaded = false;
-      this.options = [];
+      if (this.ui("dynamic")) {
```
Comment (devmiguelangel) [inline resources/js/admin/settings/components/SettingCheckboxes.vue:125]: use single quotes

```suggestion
      if (this.ui('dynamic')) {
```

File: resources/js/admin/settings/components/SettingCheckboxes.vue [hunk 6]
```diff
@@ -112,8 +122,12 @@ export default {
...
+        this.loaded = false;
+        this.options = [];
+      }
+      this.invalid = false;
+      this.message = "";
```
Comment (devmiguelangel) [inline resources/js/admin/settings/components/SettingCheckboxes.vue:130]: use single quotes

```suggestion
      this.message = '';
```

File: resources/js/admin/settings/components/SettingCheckboxes.vue [hunk 7]
```diff
@@ -129,8 +143,19 @@ export default {
         });
       }
     },
-    onSave() {
-      this.input = this.copy(this.transformed);
+    async onSave() {
+      const testSettingEndpoint = this.ui("testSettingEndpoint");
```
Comment (devmiguelangel) [inline resources/js/admin/settings/components/SettingCheckboxes.vue:147]: use single quotes

```suggestion
      const testSettingEndpoint = this.ui('testSettingEndpoint');
```

File: ProcessMaker/Http/Controllers/Auth/TwoFactorAuthController.php [hunk 8]
```diff
@@ -124,4 +128,77 @@ public function displayAuthAppQr(Request $request)
...
+
+        return response()->json([
+            'status' => 'success',
+            'message' => ('Configuration tested successfully.'),
+        ]);
+    }
+
+    private function testEmailServer()
+    {
+
```
Comment (devmiguelangel) [inline ProcessMaker/Http/Controllers/Auth/TwoFactorAuthController.php:159]: Remove line
```suggestion
```

File: ProcessMaker/Http/Controllers/Auth/TwoFactorAuthController.php [hunk 9]
```diff
@@ -124,4 +128,77 @@ public function displayAuthAppQr(Request $request)
...
+            // Build body
+            $body = $user->username . PHP_EOL . PHP_EOL;
+            $body .= __('This is a test') . PHP_EOL . PHP_EOL;
+
+            // Send SMS using Twilio SDK
+            $twilio = new Client($sid, $token);
+            $twilio->messages->create($to,
+                [
+                    'from' => $from,
+                    'body' => $body
```
Comment (devmiguelangel) [inline ProcessMaker/Http/Controllers/Auth/TwoFactorAuthController.php:193]: add comma
```suggestion
                    'body' => $body,
```

File: ProcessMaker/Http/Controllers/Auth/TwoFactorAuthController.php [hunk 10]
```diff
@@ -124,4 +128,77 @@ public function displayAuthAppQr(Request $request)
...
+                ]
```
Comment (devmiguelangel) [inline ProcessMaker/Http/Controllers/Auth/TwoFactorAuthController.php:194]: add comma
```suggestion
                ],
```

**Review by devmiguelangel — CHANGES_REQUESTED** [review]
@caleeli I put some comments

**Review by devmiguelangel — APPROVED** [review]

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-cd0b46d888.eng.processmaker.net

**Pull request merged**